### DISCUSSION TITLE
Difficulties in getting started with Imported Mode and Jupyter

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#14] [Q&A] [answered]
Hi,

I'm reading the tutorials, I found the tutorials may have a gap in terms of instructions on how to use Imported Mode. From reading other parts of the documentation I understood Imported Mode at this point works only on Jupyter notebooks. 

You might want to investigate how Bernardo, first wih transcrypt and now following Luca Damasco's Pyodide proof of concept, is working on github.com/berinhard/pyp5js with the constants and "system variables" in the global namespace ...

Also, do you think we could manage a Google Collaboratory notebook or a http://MyBinder.org link with the environment ready for teaching?

Cheers,
Alexandre

**Comment by hx2A** [comment, marked as answer]
> I found the tutorials may have a gap in terms of instructions on how to use Imported Mode. From reading other parts of the documentation I understood Imported Mode at this point works only on Jupyter notebooks.

Now that you've installed the kernel, is imported mode working for you?

For Mac users (you use a Mac?) py5 needs to be run using Jupyter because you need the `%gui osx` magic. See the [Special Notes for Mac Users](http://py5.ixora.io/tutorials/mac-users/) for more info.

For everyone else, there's also a command line tool called `run_sketch` that takes a file containing imported mode code as an argument. I haven't documented that yet though, and don't think it is that convenient. Using Jupyter is probably the best option.

> You might want to investigate how Bernardo, first wih transcrypt and now following Luca Damasco's Pyodide proof of concept, is working on github.com/berinhard/pyp5js with the constants and "system variables" in the global namespace ...

Hmmm, I hadn't heard of Pyodide, and have not looked at the pyp5js source code. It looks like this is taking a totally different approach from what I am doing here.

The hardest part about py5's imported mode was getting the dynamic variables like `mouse_x` and `mouse_y` to work correctly. The py5 kernel is identical to the regular Python kernel except it does an AST transformation before executing the byte code. It took many iterations before I arrived at something that worked.

> Also, do you think we could manage a Google Collaboratory notebook or a http://MyBinder.org link with the environment ready for teaching?

This is at the top of my priority list. Stay tuned...

**Reply by villares** [reply]
Wonderful!
![Peek 29-05-2021 18-01](https://user-images.githubusercontent.com/3694604/120084652-3d467b00-c0a8-11eb-81be-bec0f833387e.gif)

**Reply by hx2A** [reply]
Excellent!! Very happy to see this.

**Reply by hx2A** [reply]
> This is at the top of my priority list. Stay tuned...

I made a lot of progress on this today and will have a prototype to show you by the end of the week.

**Reply by hx2A** [reply]
Actually, it might not work so well, but I now know it is at least possible for some kind of online setup to work. mybinder might have some limitations that I have to deal with.

**Reply by villares** [reply]
Oh, FYI, people from the github.com/shoebot project are **really nice** you could ask them for tips with mybinder https://notebooks.gesis.org/binder/v2/gh/shoebot/jupyter-shoebot/1178e5c6fda63b3ee9f4c22d3e17e367f81eb7a8
(we could say shoebot is a lovely Processing Python relative of the drawbot/nodebox 2D branch of the family)

**Reply by hx2A** [reply]
Fascinating. I've never heard of shoebot but I know of drawbot. Allison Parish told me about it (who you might know from Processing Python) and suggested I think about how I can add functionality to py5 to allow users to make quick, simple images. The IPython magics and drawing helper tools address this, but perhaps there is more I can do to support those goals.

Thanks for that shoebot mybinder link. I just looked at how it worked and I realized they are using similar configuration features that py5 will need  to use. Processing is complicated in that it needs the JVM and a frame buffer to run. I need to install extra linux packages and start xvfb for it to work. 

https://github.com/processing/processing/wiki/Running-without-a-Display

And I will definitely reach out to them if I have questions! How do you know them? I do realize I need to learn how to be more involved on social media in the tech space, both for this project and for my own career goals.

**Reply by villares** [reply]
I think I met Ricardo on Twitter (https://twitter.com/rlaf), but he is active on the https://libregraphicsmeeting.org/ community, and I have other friends in common, friends from hackerspaces around the world :)

You can reach them at https://app.element.io/    +   #/room/#shoebot:matrix.org

**Reply by hx2A** [reply]
That was easy:

https://github.com/hx2A/py5examples
https://mybinder.org/v2/gh/hx2A/py5examples/HEAD?urlpath=lab

Over the past week my coding efforts all related to running py5 online through mybinder or JupyterHub. I found and fixed a critical bug and made some enhancements that will greatly improve the usefulness of this.

I'll do another release by this weekend and create some worthwhile example code for the py5examples repo.

**Reply by villares** [reply]
woderful!

**Reply by hx2A** [reply]
@villares , when you have a moment, run this notebook through mybinder:

https://mybinder.org/v2/gh/hx2A/py5examples/HEAD?filepath=examples%2FRotating%20Cube.ipynb

It is using a dev version of py5. Turns out I won't be around this weekend, but I'll do a release next week for sure.

**Reply by hx2A** [reply]
Also, I just emailed some Processing Foundation people about how to introduce py5 to the larger community.

**Reply by villares** [reply]
![image](https://user-images.githubusercontent.com/3694604/121775850-42073680-cb60-11eb-97a8-475dbc8af35a.png)
Very nice!
